src/Xenial.Tasty.Cli/Program.cs

```diff
@@ -0,0 +1,236 @@
+                    {
+                        Console.WriteLine($"{c.Name} - {c.Description}" + (c.IsDefault ? " (default)" : string.Empty));
+                    }
+
+                    Console.WriteLine("c - Cancel");
+                    Console.WriteLine("================");
+                };
+
+                Func<ConsoleKeyInfo, SerializableTastyCommand?> findCommand = (info) =>
+                {
+                    var key = info.Key.ToString().ToLower();
+
+                    var command = info.Key == ConsoleKey.Enter
+                        ? commands.FirstOrDefault(c => c.IsDefault)
+                        : commands.FirstOrDefault(c => c.Name.ToLowerInvariant() == key);
+
+                    return command;
+                };
+
+                Func<Task<Func<Task>>> readInput = async () =>
+                {
+                    var info = await readConsoleKey();
+                    var command = findCommand(info);
+                    if (command != null)
+                    {
+                        return async () =>
+                        {
+                            Console.WriteLine($"Executing {command.Name} - {command.Description}");
+
+                            await tastyServer.DoExecuteCommand(new ExecuteCommandEventArgs
+                            {
+                                CommandName = command.Name
+                            });
+                        };
+                    }
+                    if (info.Key == ConsoleKey.C)
+                    {
+                        return () => Promise(async (_, reject) =>
+                        {
+                            Console.WriteLine($"Requesting cancellation");
+
+                            await tastyServer.DoRequestCancellation();
+                            reject();
+                        });
+                    }
+
+                    return () =>
+                    {
+                        return Task.CompletedTask;
+                    };
+                };
+
+                Task waitForInput() => Promise(async (resolve, reject) =>
+                {
+                    writeCommands();
+                    var input = await readInput();
+
+                    var stop = Promise(async (_, reject) =>
+                    {
+                        var cancel = cancelKeyPress();
+                        var endTestPipeline = endTestPipelineSignaled();
+                        try
+                        {
+                            await Task.WhenAny(cancel, endTestPipeline);
+                        }
+                        catch (TaskCanceledException)
+                        {
+                            reject();
+                        }
+                    });
+                    var inputTask = input();
+                    var tasks = new[] { inputTask, stop };
+                    await Task.WhenAny(tasks);
+                    if (!tasks.Any(m => m.Status == TaskStatus.Canceled))
```

Comment:
This doesn't work as expected yet. We need to find a way to signal back from the client the actual state and raise an cancellation request from the CLI back to the client, without messing up totally with the simple client architecture. But we totally don't have a notion of cancellation there yet though...